### PR TITLE
PHOENIX-6165 Add Phoenix-Connectors assembly to create tarball of jar…

### PR DESCRIPTION
**Pull request by richardantal**
… artifacts

### CODE REVIEW

File: phoenix-connectors-assembly/src/build/components/phoenix4-jars.xml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+
+-->
+<component>
```
Comment (stoty) [inline phoenix-connectors-assembly/src/build/components/phoenix4-jars.xml:22]: I think that this would be much cleaner using dependency sets.

File: phoenix-connectors-assembly/src/build/package-phoenix4-connectors-to-tar-all.xml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  <!-- Components that we don't want in jars that are used with other libraries, but we want for a standalone client -->
+  <dependencySets>
+    <dependencySet>
+      <!-- Unpack all the dependencies to class files, since java doesn't support
+        jar of jars for running -->
+      <unpack>false</unpack>
+      <!-- save these dependencies to the top-level -->
+      <outputDirectory>/lib</outputDirectory>
+      <includes>
+        <include>org.slf4j:slf4j-log4j12</include>
```
Comment (stoty) [inline phoenix-connectors-assembly/src/build/package-phoenix4-connectors-to-tar-all.xml:42]: Why is this necessary ?

File: pom.xml [hunk 3]
```diff
@@ -104,13 +105,16 @@
...
+    <skip-phoenix5-assembly>false</skip-phoenix5-assembly>
+    <skip-phoenix4-assembly>false</skip-phoenix4-assembly>
 
     <!-- Set default encoding so multi-byte tests work correctly on the Mac -->
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
     <curator.version>2.12.0</curator.version>
 
     <shaded.package>org.apache.phoenix.shaded</shaded.package>
+    <sqlline.version>1.9.0</sqlline.version>
```
Comment (stoty) [inline pom.xml:117]: Why is sqlline needed for the connectors ?

File: pom.xml [hunk 4]
```diff
@@ -104,13 +105,16 @@
     <numForkedIT>7</numForkedIT>
     <it.failIfNoSpecifiedTests>false</it.failIfNoSpecifiedTests>
     <surefire.failIfNoSpecifiedTests>false</surefire.failIfNoSpecifiedTests>
+    <skip-phoenix5-assembly>false</skip-phoenix5-assembly>
+    <skip-phoenix4-assembly>false</skip-phoenix4-assembly>
```
Comment (stoty) [inline pom.xml:109]: Come to think of it, it would save build/test time if we could also skip building the phoenix4/5 modules alltogether.

File: phoenix-connectors-assembly/src/build/package-phoenix4-connectors-to-tar-all.xml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      </includes>
+    </dependencySet>
+  </dependencySets>
+
+  <componentDescriptors>
```
Comment (stoty) [inline phoenix-connectors-assembly/src/build/package-phoenix4-connectors-to-tar-all.xml:47]: Not sure where that would go exactly, but adding symlinks without the (connectors) version in the filename would be helpful.

**Review by stoty — CHANGES_REQUESTED** [review]

File: phoenix-connectors-assembly/pom.xml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            <groupId>org.apache.phoenix</groupId>
+            <artifactId>phoenix4-pig</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.phoenix</groupId>
+            <artifactId>phoenix5-pig</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (richardantal) [reply phoenix-connectors-assembly/pom.xml:86]: No, probably it isn't.

File: phoenix4-connectors-assembly/pom.xml [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                            <executable>ln</executable>
+                            <workingDirectory>${project.basedir}/../phoenix-hive-base/phoenix4-hive/target</workingDirectory>
+                            <arguments>
+                                <argument>-fnsv</argument>
+                                <argument>
+                                    phoenix4-hive-${project.version}.jar
+                                </argument>
+                                <argument>
+                                    <!-- We are overwriting the unshaded server JAR, but we don't care -->
+                                    phoenix-hive.jar
```
Comment (stoty) [inline phoenix4-connectors-assembly/pom.xml:137]: I am not sure if removing phoenix4 / phoenix5 from the name is a good idea. I lean towards keeping that part. (same applies to every symlink)

**Review by stoty — CHANGES_REQUESTED** [review]
Getting closer

File: pom.xml (hunk 4 above)
Comment (richardantal) [reply pom.xml]: That is a great idea!

File: phoenix4-connectors-assembly/pom.xml [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+                                </argument>
+                                <argument>
+                                    phoenix4-kafka.jar
+                                </argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
```
Comment (stoty) [inline phoenix4-connectors-assembly/pom.xml:202]: The directory created by this section is not copied into the assembly file.
Luckily, since if it was, there would be nothing to serve the artifacts from there.

Please remove this (from both assemblies), or explain why this should be kept.

**Review by stoty — CHANGES_REQUESTED** [review]

File: phoenix5-connectors-assembly/pom.xml [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                                </argument>
+                                <argument>
+                                    phoenix5-kafka.jar
+                                </argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
```
Comment (stoty) [inline phoenix5-connectors-assembly/pom.xml]: What purpose does this plugin serve ?

File: phoenix5-connectors-assembly/pom.xml [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        <groupId>org.apache.phoenix</groupId>
+        <artifactId>phoenix-connectors</artifactId>
+        <version>6.0.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>phoenix5-connectors-assembly</artifactId>
+    <packaging>pom</packaging>
+    <name>Phoenix 5 Connectors Distribution Assembly</name>
+
+    <profiles>
```
Comment (stoty) [inline phoenix5-connectors-assembly/pom.xml:35]: The profile selection is already handled in the root pom, this can be removed.

File: phoenix4-connectors-assembly/pom.xml [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        <groupId>org.apache.phoenix</groupId>
+        <artifactId>phoenix-connectors</artifactId>
+        <version>6.0.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>phoenix4-connectors-assembly</artifactId>
+    <packaging>pom</packaging>
+    <name>Phoenix 4 Connectors Distribution Assembly</name>
+
+    <profiles>
```
Comment (stoty) [inline phoenix4-connectors-assembly/pom.xml:35]: I don't think profiles are needed here, as the module shouldn't even be run anymore.

**Review by stoty — CHANGES_REQUESTED** [review]
The real problem that I have just noticed is that you are copying the unshaded JARs, which are not very useful.
Users need the shaded artifacts for each connector.

**Comment by stoty** [comment]
All that remains to be done is packaging the shaded connectors instead.

**Review by stoty — APPROVED** [review]
LGTM
